app/routes/blog.$slug.test.tsx

```diff
@@ -1,32 +1,84 @@
 // @vitest-environment node
 import "@testing-library/jest-dom";
 import { mockUrlResponse, urlToLoaderArgs } from "../../test/test-utils";
-import { loader } from "~/routes/blog.$slug";
+import { loader, longCacheMaxAge, shortCacheMaxAge } from "~/routes/blog.$slug";
 import { bundleMdx } from "../../blog/parser";
 import { POSTS_BUILD_DIR } from "blog/paths";
 
 const TEST_URL = "https://test.io";
 const SLUG = "slug";
 const TEST_GET_BLOG_POST_URL = `${TEST_URL}/resource/get-blog-post/${SLUG}`;
 const TEST_POST_URL = `${TEST_URL}/${SLUG}`;
-const TEST_MDX = `---
+const ONE_MINUTE_MS = 1000 * 60;
+const ONE_DAY_MS = ONE_MINUTE_MS * 60 * 24;
+
+describe("/blog/$slug", () => {
+  it("should bundle mdx", async () => {
+    const todayPublishDate = new Date().toISOString();
+    const testMdx = `---
 title: Title
+date: ${todayPublishDate}
 ---
 # Header
 
 Hello, world.
 `;
-
-describe("/blog/$slug", () => {
-  it("should bundle mdx", async () => {
     mockUrlResponse(TEST_GET_BLOG_POST_URL, {
-      json: await bundleMdx(TEST_MDX, SLUG, POSTS_BUILD_DIR),
+      json: await bundleMdx(testMdx, SLUG, POSTS_BUILD_DIR),
     });
     const args = urlToLoaderArgs(TEST_POST_URL, { path: { slug: SLUG } });
+    const response = await loader(args);
     const {
       post: { code, ...metaProps },
-    } = await loader(args);
-    expect(metaProps).toEqual({ slug: SLUG, frontmatter: { title: "Title" } });
+    } = await response.json();
+    expect(metaProps).toEqual({
+      slug: SLUG,
+      frontmatter: { title: "Title", date: todayPublishDate },
+    });
     expect(code).toContain("Hello, world.");
   });
+
+  it("should have a Cache-Control header with a max-age of 8 hours if post is less than 4 days old", async () => {
+    const today = Date.now();
+    const fourDaysAgoMs = today - ONE_DAY_MS * 4 + ONE_MINUTE_MS; // 1 minute "safety buffer"
+    const publishDate = new Date(fourDaysAgoMs).toISOString();
+    const testMdx = `---
+title: Title
+date: ${publishDate}
+---
+# Header
+
+Hello, world.
+`;
+    mockUrlResponse(TEST_GET_BLOG_POST_URL, {
+      json: await bundleMdx(testMdx, SLUG, POSTS_BUILD_DIR),
+    });
+    const args = urlToLoaderArgs(TEST_POST_URL, { path: { slug: SLUG } });
+    const response = await loader(args);
+    const cacheControlHeader = response.headers.get("Cache-Control");
+    expect(cacheControlHeader).not.toBeNull();
+    expect(cacheControlHeader).toBe(shortCacheMaxAge);
+  });
+
+  it("should have a Cache-Control header with a max-age of 1 week if post is more than 4 days old", async () => {
+    const today = Date.now();
+    const fourDaysAgoMs = today - ONE_DAY_MS * 4 - ONE_MINUTE_MS; // 1 minute "safety buffer"
```

Comment:
instead of a safety buffer, look into [vitest's date mocking](https://vitest.dev/guide/mocking.html#dates). you can likely get rid of a `Date.now` reference and be explicit in the test. and then you can just hardcode the date in the test MDX that way you don't have to calculate it which will make it a bit more simple to read.

Reply:
I'm glad I know about this now. At my job, we used to have some time-based tests that would fail because `Date.now()` would sometimes run one millisecond apart between test code and functional code. Glad we can lock in a firm value for `Date.now()`.

I made a change that uses `vi.setSystemTime(date)`. I didn't remove usages of `Date.now()`. But I did remove the buffers.

Reply:
```typescript
      vi.setSystemTime(Date.now());
      const fourDaysAgoMs = Date.now() - ONE_DAY_MS * 4;
```
this still _technically_ has the same issue since `Date.now` is called twice. it could theoretically return two diff timestamps if the timing is just right (or just wrong depending on your perspective). it's unlikely, but enough runs of this and that edge case will mysteriously and transiently pop up, giving you a heisenbug. best practice here is something along the lines of

```typescript
     const now = new Date('2023-02-03T02:02:34.000Z')
      vi.setSystemTime(now);
      const fourDaysAgoMs = now - ONE_DAY_MS * 4;
```

i usually put my `now` as a constant at the top of the file and `setSystemTime` in a `beforeEach` so every test has the same timestamp. a bit easier to reason about.

Reply:
This isn't the case when `vi.useFakeTimers` is called. When that is called in `beforeEach` and combined with `vi.setSystemTime()` all calls to time based functions, including `Date.now`, are wrapped and locked in place. They should never return different values no matter how far apart they are called.

The reason `vi.useFakeTimers` is used is because my loader function also calls `Date.now()`. So even if I remove all usages of `Date.now()` in the test, I still have one used in the implementation. I have to lock time in place for the whole runtime.

You can check [this Stackblitz demo](https://stackblitz.com/edit/node-oxx7ga?file=test%2Fbasic.test.ts) to see what I mean.

Reply:
After taking a moment to chew this over, I think I have a way to clean this up that should make us both happy. I'll push it up now.

While `vi.useFakeTimers()` does do what I said, in that it locks `Date.now()` to a constant return value, I do think that seeing it called like I did in the test could be confusing for newcomers. So I'll remove usages of that in the test. But I do still need to use `vi.useFakeTimers()` to lock time in place and make sure that the call to `Date.now()` in my loader is the same value as the one passed `vi.setSystemTime()`.

Reply:
> all calls to time based functions, including Date.now, are wrapped and locked in place. 

well damn. learn something new everyday. i didn't realize that. awesome. 

i'm fine w/ you keeping what you had then. chalk that up to me knowing enough about the api to be dangerous 😅